### PR TITLE
Refactor ansible

### LINKED ISSUE #59: Change MediaWiki configurations to have multiple files
Organically the system has grown where all the configuration gets stuck in LocalSettings.php from ansible `lineinfile` and `blockinfile` tasks.  This is the normal way to configure MediaWiki, however it makes both the ansible files, and LocalSettings a bit unwiedly.

Instead, we should have some loop in LocalSettings that loads up all the files in a directory, not unlike other configuration schemes in linux, and those files being part of the files/templates folder in ansible, rather than using `blockinfile` in the tasks yml file.

This is related to #56, as the general LocalSettings for the entire box could just loop over a directory that's symlinked for the given competition.  So it makes sense to do these two issues at the same time.

### PR DESCRIPTION
**Pull request by frankduncan**
Wow, 6100 lines of code removed!  This fixes #59, #62, and #64, in a formal way, in various commits in the chain.

Go ahead and review as you like, though I'll be merging in the next few days before I upgrade mediawiki and push this all to production.

### CODE REVIEW

File: roles/mediawiki/files/lfc-analysis-rule.php [hunk 1]
```diff
@@ -0,0 +1,41 @@
+<?php
```
Comment (kfogel) [inline roles/mediawiki/files/lfc-analysis-rule.php:1]: Some comment somewhere near the top of this file, explaining what it's about / what it's for, would be good.  Most people won't see "lfc-analysis" and automatically associate that with "pages that are conditionally visible and may have UI changes designed to make them behave differently from main proposal pages".

File: roles/mediawiki/files/special-log.php [hunk 2]
```diff
@@ -0,0 +1,46 @@
+<?php
+
+# Make it so names of log types are more amenable to LFC users
```
Comment (kfogel) [inline roles/mediawiki/files/special-log.php:3]: Well, more specifically, to LFC *staff*.

File: roles/mediawiki/files/special-log.php [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+
+$wgHooks['MessagesPreLoad'][] = function($title, &$message, $code) {
+  global $logTitlesOverrideMessages;
+  if(array_key_exists(strtolower($title), $logTitlesOverrideMessages)) {
+    $message = $logTitlesOverrideMessages[strtolower($title)];
+  }
+};
+
+
+# Remove log types that we don't want
```
Comment (kfogel) [inline roles/mediawiki/files/special-log.php:27]: Yeah -- the comments in this file are really helpful.  This kind of commenting would be good to have too in `lfc-analysis-rule.php`.

File: roles/permissions/files/default-permissions.php [hunk 4]
```diff
@@ -0,0 +1,85 @@
+<?php
+# Even if these groups aren't used in the wiki (no users assigned), this should be the
```
Comment (kfogel) [inline roles/permissions/files/default-permissions.php:2]: Very useful comment!

File: roles/picksome/files/picksome.php [hunk 5]
```diff
@@ -0,0 +1,67 @@
+<?php
+# While this next line should be taken care of in extenions, and indeed
```
Comment (kfogel) [inline roles/picksome/files/picksome.php:2]: I found this sentence somewhat hard to parse :-).

File: roles/picksome/files/picksome.php [hunk 6]
```diff
@@ -0,0 +1,67 @@
...
+# it is, because this extension is loaded via wfLoadExtension, and because
+# Collection is loaded the old way, the sidebar before output is called
+# preferring Collection, placing it higher on the siderbar than PickSome.
+#
+# This was requested to be changed, and the most straightforward way is
```
Comment (kfogel) [inline roles/picksome/files/picksome.php:7]: I think it would be worth being more verbose here, for the reader who doesn't have as much context as you do.

File: roles/picksome/files/picksome.php [hunk 7]
```diff
@@ -0,0 +1,67 @@
...
+# to add it to the hook directly when loaded by LocalSettings
+$wgHooks['SidebarBeforeOutput'][] = 'PickSomeHooks::onSidebarBeforeOutput';
+wfLoadExtension('PickSome');
+$wgPickSomeNumberOfPicks = 15;
+
+$wgLFCPickSomeEligiblePage = "TorqueConfig:ValidProposals";
+
+# This may not be performant and require caching if number of users/page hits
```
Comment (kfogel) [inline roles/picksome/files/picksome.php:15]: "This may not be performant, and it may require caching if number of users/page hits becomes large."

File: roles/picksome/files/picksome.php [hunk 8]
```diff
@@ -0,0 +1,67 @@
...
+$wgPickSomeSortFunction = function($t1, $t2) {
+  $text1 = $t1->getText();
+  $text2 = $t2->getText();
+
+  $text1 = preg_replace("/^\\W/", "", $text1);
+  $text2 = preg_replace("/^\\W/", "", $text2);
+  return $text1 > $text2;
+};
+
+# These are defaults for competitions, and probably will be overridden at some point
```
Comment (kfogel) [inline roles/picksome/files/picksome.php:44]: When you say "probably will be overridden at some point", you mean "probably will be overridden by competition-specific settings within a given competition", right?

File: roles/simplefavorites/tasks/main.yml [hunk 9]
```diff
@@ -0,0 +1,12 @@
+---
+
+- name: Checkout SimpleFavorites
```
Comment (kfogel) [inline roles/simplefavorites/tasks/main.yml:3]: Hill.  Dying on.

File: roles/simplesaml/templates/simplesaml.php.j2 [hunk 10]
```diff
@@ -0,0 +1,82 @@
+<?php
+
+# When including this file, a competition needs to override $wgPluggableAuth_EnableLocalLogin
+# before this file is loaded.  That's because the file takes care of the local login code
```
Comment (kfogel) [inline roles/simplesaml/templates/simplesaml.php.j2:4]: "because this file"

File: roles/teamcomments/tasks/main.yml [hunk 11]
```diff
@@ -0,0 +1,19 @@
+---
+
+- name: Checkout TeamComments
```
Comment (kfogel) [inline roles/teamcomments/tasks/main.yml:3]: Not saying anything more about hills.  Maybe I'm just receiving medical treatment on them, and there is still hope.

**Review by kfogel — APPROVED** [review]

File: 100Change2020/ansible/100Change2020.yml [hunk 12]
```diff
@@ -16,4 +16,19 @@
...
   become_user: "{{ deployment_user }}"
   roles:
-  - mediawiki
+    - mediawiki
+    - activitylog
+    - collection
+    - embed_video
+    - permissions
+    - simplefavorites
+    # PickSome has to come after collection so the left menu bar is correct
```
[the remark this answers is not in the document]
Comment (frankduncan) [reply 100Change2020/ansible/100Change2020.yml:25]: I can add something to our Collection config that will error if PickSome is already installed.  I'll put it on my todo list, because I don't think it needs to happen in this PR.

File: 100Change2020/ansible/inv/dev/group_vars/all/base [hunk 13]
```diff
@@ -1,4 +1,6 @@
 ---
+competition_name: lfc
```
[the remark this answers is not in the document]
Comment (frankduncan) [reply 100Change2020/ansible/inv/dev/group_vars/all/base:2]: base is the configuration that can get checked in for production, so this doesn't get overridden later, as this is IS the overriding configuration.  "secret" is passwords and whatnot that shouldn't be.  See https://github.com/OpenTechStrategies/torque-sites/blob/main/INSTALL.md#production-installs-for-ots-only for more information.

File: 100Change2020/ansible/roles/100Change2020/tasks/main.yml [hunk 14]
```diff
@@ -0,0 +1,46 @@
+---
+
+# Because the overrides MUST come after the originals in LocalSettings, we
```
[the remark this answers is not in the document]
Comment (frankduncan) [reply 100Change2020/ansible/roles/100Change2020/tasks/main.yml:3]: Because no one should be editing the local settings file, and the comment talks about ansible things (deleting the block), the comment should more appropriately here because the decisions are all made here.  I did update it to make t more clear.

File: roles/activitylog/tasks/main.yml [hunk 15]
```diff
@@ -0,0 +1,24 @@
+---
+
+- name: Checkout ActivityLog
```
[the remark this answers is not in the document]
Comment (frankduncan) [reply roles/activitylog/tasks/main.yml:3]: Something something fluidity of language something something.

File: roles/finishmediawiki/tasks/main.yml [hunk 16]
```diff
@@ -0,0 +1,12 @@
+---
+
+# This should come near the end to make sure that all the database updates
+# for the extensions get run after the rest of the system is set up.
+- name: Run mediawiki updates
+  command: "php {{ mediawiki_install_directory }}/mediawiki-1.33.0/maintenance/update.php"
```
[the remark this answers is not in the document]
Comment (frankduncan) [reply roles/finishmediawiki/tasks/main.yml:6]: We are, but I wanted to separate the updating stuff from the refactor.

File: roles/picksome/files/picksome.php (hunk 5 above)
Comment (frankduncan) [reply roles/picksome/files/picksome.php:2]: I cleaned this up, and the following too.

File: roles/picksome/files/picksome.php (hunk 8 above)
Comment (frankduncan) [reply roles/picksome/files/picksome.php:44]: Yes, exactly.  Updated to be more clear.

File: roles/teamcomments/tasks/main.yml (hunk 11 above)
Comment (frankduncan) [reply roles/teamcomments/tasks/main.yml:3]: I really tempted to leave this one, just because you were less committed to it than the others.

**Comment by frankduncan** [comment]
Thanks so much for the review!  I updated many things and am merging, but feel free to ping me about anything that's still unclear. (I need to merge so I can get the next competition loaded in)
